[client] Add client metadata header on RestClient requests

swallez opened the pull request:
Supersedes #66189 to target `master`. See there for previous review comments.

Adds a `X-Elastic-Client-Meta` header to http requests sent by `RestClient`. This header contains information about the runtime environment that is meant to allow analyzing usage context by collecting this information on the receiving side of requests, like a proxy server in front of ES.

The information in this header has been specified by the clients team to be consistent across language clients. It contains the versions of the client, the language runtime (Java in this case) and its http library, and the use of the high level client. Additional JVM specific information is added about standard libraries for other JVM languages like Kotlin, Scala, Groovy and Clojure, so that we can extrapolate interest for dedicated clients for these languages.

This header is sent by default and can be disabled using `RestClientBuilder.setMetadataHeaderEnabled(false)`.

This PR also sets the `User-Agent` header to `elasticsearch-java/<version> (Java/<version>)` to again be consistent with other language clients. Users are however free to override this value with application-specific agent information.

swallez committed:
[client] Add client metadata header on RestClient requests

swallez committed:
Change Throwable to Exception, add test

swallez committed:
Rename 'v' to 'version'

swallez committed:
Read version from Manifest to avoid hardcoding the version in the client

swallez committed:
Fix style issue & forbidden API

swallez committed:
Compute header values in static initializer

swallez committed:
Fix combined runtime info calculation

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

sethmlarson commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+            s.append(",hl=").append(version);
+        }
+
+        version= kotlinVersion();
```
Do any of these version functions need to handle pre-release information?

swallez replied:
Service/language/http-client versions are not stripped, and so can include pre-release information. For language runtimes we only keep major.minor as this information allows us to better understand the ecosystem and the patch version doesn't bring useful information. The patch version for these may only be of interest if we build dedicated clients for these languages, in which case their version will be reported as the language version.

sethmlarson commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+        return null;
+    }
+
+    static String stripPatchRevision(String version) {
```
If patch numbers aren't long can we keep this info anyways?

swallez replied:
They can be long depending on the language, like `2.13.0-M5-6e0cba7` for some release of Scala. We could also compact pre-release versions but there's honestly no real value in the patch revision for this particular use case where we just want to know if a particular language is present, and in what major version (minor language versions are often not exactly minor). The patch number won't give us any actionable information unless we have a dedicated client for that language, in which case the full version will be reported in the first part of the header.

sethmlarson commented on client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java:
```diff
@@ -56,6 +68,39 @@
     private NodeSelector nodeSelector = NodeSelector.ANY;
     private boolean strictDeprecationMode = false;
     private boolean compressionEnabled = false;
+    private boolean metadataHeaderEnabled = true;
```
metaHeaderEnabled to conform with the `meta header` noun

swallez replied:
Done. Also renamed constants.

sethmlarson reviewed (changes requested):
Looks good, some comments for you!

rjernst commented on client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java:
```diff
@@ -220,11 +276,20 @@ private CloseableHttpAsyncClient createHttpClient() {
+            // Always add metadata header last so that it's not overwritten
+            httpClientBuilder.addInterceptorLast((HttpRequest request, HttpContext context) -> {
+                if (metadataHeaderEnabled) {
+                    request.setHeader(METADATA_HEADER, METADATA_HEADER_VALUE);
```
While this is better, I still think we should not be silently allowing leniency in the computation. If the header is enabled, we should have a valid value, or throw an error. My original suggestion was to lazily set the cached header value (it could and should be outside this lambda, since as you say, the enabled flag is per client, we don't need to check it on every request). This way the method computing the header value does not need all of the try/catch for failures (with one exception: we still need to handle the various possible runtimes, but we should narrow those to ClassNotFoundException).

swallez replied:
The header was already computed outside the lambda. Moving it to a static initializer allows computing it only once instead of at every construction of a client.

Regarding exceptions and leniency, this header is enabled by default and should not be something users should be concerned with, except if they want to opt out telemetry by calling `RestClientBuilder.setMetadataHeaderEnabled(false)`. That's all. In particular we should not require users to change their build system if it strips out the manifests we read the version from.

No error related to this header should ever bubble up to the user, and we can accept it to contain partial information: knowing that an ES Java client is making a request even if we could not find its version is already interesting information as it contributes to aggregated statistics about the Java client. Since we know it can happen, the aggregation logic will take care of it, and so in this context partial information is still valid information.

And we can then work to understand why the version was not found and fix/update the discovery code.

Regarding runtimes, we also have to be robust against errors that go beyond the existence of the class. For example an application may have a language runtime in a older version that does have the class we're checking, but not the field/method we read the version from. We should not fail in that case.

We should actually return partial information in that case too (e.g. "I found Groovy, but couldn't find its version") which this PR currently does not do. I have updated my [playground for runtime sniffing](https://github.com/swallez/jvm-languages-versions/blob/2bc92312a84f608b9e87328e6a9262edda37adb3/src/main/java/RuntimeInfo.java) for this.

WDYT? Does this make sense? Did I miss something?

rjernst replied:
Thanks for additional explanation. I understand the desire to not cause an error for a user, since this is something internal. It's understandable for the runtime. But for the Elasticsearch version or apache version, which make available through the build of our client, it could mask a bug on our end. Of course so could the runtimes; we could have a typo in the class or method names we introspect.

The thing that gives me pause about all this code is there are not really tests for the actual data, only for "was anything written in the response?"  Without tests for each of these runtimes, we can't be sure the code actually works. And for Elasticsearch version, using a fatjar seems like something that may be common, so I think we should expect any users building such a jar to include the appropriate manifest value. However, it is certainly something we can add as a requirement later, since something is better than nothing right now.

swallez replied:
> Without tests for each of these runtimes, we can't be sure the code actually works.

That's a fair point, and is why I have this "[playground project](https://github.com/swallez/jvm-languages-versions)" to test various runtimes. I see two possible approaches here:
- add all those runtimes as _test_ dependencies of the rest client.
- have a separate project like the one above where we can test more language runtime variants, and have the rest client's `RuntimeInfo` code be a direct copy of the one tested in that separate project.

First approach keeps the ES repo self-contained, but adds some dependencies that, even if limited to tests, add some weight to the build. It also doesn't allow testing different versions of a given language runtime unless we create additional Gradle subprojects.

Second approach allows having a more complex build with a main project containing the `RuntimeInfo` code, and several subprojects for each runtime version we want to test (e.g. scala-2.12, scala-2.13, scala-3). This allows testing `RuntimeInfo` with many combinations.

The drawback of this second approach is the manual operation to copy updated `RuntimeInfo` from that project to the rest client's code base, unless we can somehow automate it.

> it is certainly something we can add as a requirement later, since something is better than nothing right now.

Exactly. This is why I initially hard-coded the version as a string, but the additional maintenance burden that you pointed out led me to use what is available now and consider improving things later, e.g. by having a class generated at build time.

We should note that httpclient uses classpath resource lookup to find its version (e.g. `/org/apache/http/nio/version.properties`). Although not as strong as a hardcoded string, it should survive in a fat jar, contrarily to manifest attributes. So maybe an approach we should consider for ES itself (and the client).

swallez committed:
Rename "metadata header" to "meta header"

swallez committed:
Write version to a resource file, update language runtime lookup

swallez commented:
Added a commit with the following updates:

* the LLRC gradle build creates a resource file that contains the ES version. This approach is much more robust than a manifest entry if users create an uber-jar with the library. The previous static string that had to be kept in sync with `Version` is gone.

* discovery of alternate JVM runtime in the `LanguageRuntimeVersions` is a verbatim copy of the code in https://github.com/elastic/jvm-languages-sniffer/ that also contains tests for all supported languages. This approach avoids complexifying the Elasticsearch build with the addition of dependencies on each of those languages' runtime library while ensuring proper test coverage.

@andreidan, @rjernst can you please review?

swallez commented:
@elasticmachine run elasticsearch-ci/1 please

swallez commented:
@elasticmachine run elasticsearch-ci/2 please

swallez commented:
@elasticmachine update branch

sethmlarson reviewed (approved):
LGTM

andreidan commented on client/rest/src/main/java/org/elasticsearch/client/LanguageRuntimeVersions.java:
```diff
@@ -57,79 +54,74 @@ public static String getRuntimeMetadata() {
-            // ignore
+            clazz = Class.forName(className);
+        } catch (ClassNotFoundException e) {
+            return null;
```
Is the `return null` vs `return ""` difference purposeful? Isn't it the same "can't get the version" case?

swallez replied:
Yes, it's slightly different: we return `null` when the class doesn't exist, i.e. we're sure the language runtime isn't there (or at least the class identifying it). When it exists and we can't call the method or can't get the field, this means the information to look for has changed, and we report the language as existing but without version information. That will trigger an analysis on our end to look for updates to the language runtime.

andreidan replied:
Hm, but if `org.jruby.runtime.Constants` is moved to `org.jruby.newpackage.Constants` this invariant will not hold anymore (ie. a particular class being there or not finding for the correct class FQN are indistinguishable cases IMO). I don't have a strong opinion on this, but I'd personally go for the simpler option of "I don't know the version or if it is even present at all"

andreidan commented on client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java:
```diff
@@ -71,35 +71,38 @@
+
+        // Never fail on unknown version, even if an environment messed up their classpath enough that we can't find it.
+        // Better have incomplete telemetry than crashing user applications.
+        String version = null;
```
nit: if we initialise it here with `""` as a default value, the below `if (version == null)`check is not needed anymore

swallez replied:
Actually it is: `versions.getProperty("elasticsearch-client")` can return `null` if the property file exists but doesn't contain that property. I agree that this is unlikely to happen, but client applications are a wild environment 😛 

So I finally chose that approach so that we don't have a mix of tests for `""` and `null`.

andreidan replied:
Ah, good shout!

andreidan reviewed (approved):
LGTM, thanks for iterating on this @swallez 

I'll defer the gradle infrastructure review to @rjernst

swallez commented:
@elasticmachine update branch

swallez committed:
Get httpclient's version in a privileged context

This is needed because of the call to getClassLoader()

Fixes the failure of testThatHttpExporterAddsWatches that was caused by the test
ES serverto fail while instanciating a client to setup remote monitoring.

swallez committed:
Remove unused import

swallez commented:
@rjernst can you review the changes to the gradle build that add a resource file? I'd like to backport this to 7.12 in due time before feature freeze.

rjernst reviewed (approved):
Gradle changes look fine to me.

mark-vieira commented on client/rest/build.gradle:
```diff
@@ -43,6 +43,21 @@ dependencies {
   testImplementation "org.elasticsearch:mocksocket:${versions.mocksocket}"
 }
 
+def generateVersionResource = tasks.register("generateVersionResource") {
```
We could simplify all this by just putting a file under `src/main/resources` with contents like so:

```
elasticsearch-client=${elasticsearchVersion}
```

Then we can configure the `processResources` task to replace that value when copying the file to the resources output directory.

```
tasks.named("processResources").configure {
  expand elasticsearchVersion: versions.elasticsearch
}
```

swallez replied:
Excellent, TIL! Thanks!

I pushed 80c11f9 where I just `expand versions: versions` so that the placeholder in the properties file is the more familiar `${versions.elasticsearch}`. Can you please cross-check?

mark-vieira replied:
That'll work.

swallez committed:
Simplify properties file generation
